sdk/metric/aggregation/aggregation.go

```diff
@@ -0,0 +1,119 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//go:build go1.17
+// +build go1.17
+
+// Package aggregation contains configuration types that define the
+// aggregation operation used to summarizes recorded measurements.
+package aggregation // import "go.opentelemetry.io/otel/sdk/metric/aggregation"
+
+import (
+	"errors"
+	"fmt"
+)
+
+// Aggregation is the aggregation used to summarize recorded measurements.
+type Aggregation struct {
+	// Operation is the kind of operation performed by the aggregation and the
+	// configuration for that operation. This can be Drop, Sum, LastValue, or
+	// ExplicitBucketHistogram.
+	Operation operation
```

Comment:
Why the nesting of Operation within Aggregation?  I thought "Drop" _is_ an Aggregation?  Do we expect an aggregation to have more than an operation in the future?

Reply:
We need a way to universally refer to all aggregations (i.e. as a parameter to or return type from a function) but restrict it so there are no user allowed aggregations (something required by the OTel specification). Wrapping an unexported interface in an exported struct satisfies both of these behaviors.

If, instead, we export say:

```
type Aggregation interface {
	private()
}
```

There is nothing stopping a user from supplying:

```go
type MyAgg struct {
	aggregation.Aggregation
}

This could still ultimately be caught as a setup error and returned to the user. The goal here is to eliminate the possibility for the error entirely.

As for the naming `Operation`, before `Default` all the allowed values described the operation the aggregation would perform (defining the aggregation as applying an operation on a set from one domain and producing a set in an output domain).

That said, I'm open to suggestions (for reference, other approaches were explored in #2950).

Reply:
With the current approach, can a user just embed one of the existing types (e.g. aggregation.Drop) in their implementation of Aggregation to make a user-provided aggregation?  If so, would it be better to have something like:

```golang
type Aggregation struct {
    operation operation
}

func Drop() Aggregation {
  // the drop struct is not exported
  return Aggregation{operation: drop}
}
// ... same for other ops...
```

Reply:
> With the current approach, can a user just embed one of the existing types (e.g. aggregation.Drop) in their implementation of Aggregation to make a user-provided aggregation?

true

> If so, would it be better to have something like:
> 
> ```go
> type Aggregation struct {
>     operation operation
> }
> 
> func Drop() Aggregation {
>   // the drop struct is not exported
>   return Aggregation{operation: drop}
> }
> // ... same for other ops...
> ```

I considered having functions return an `Aggregation`, the issue is going to be future extensibility and the ability to identify the aggregation kind outside of the package.

If parameters are added to any aggregation we would need to support that in a backwards compatible manner. This leads me to think we would need options for each aggregation creation function (e.g. `func Drop(opts ...DropOption) Aggregation`). This seemed a bit excessive to me, but I'm not opposed entirely if you think its a better approach and we can solve the other issue.

When we return `Aggregatoin{operation: drop{}}` there is no way for an external reader to identify the internal state. How can we communicate this without returning a struct or other exported value? And if we are doing that we would be back where we started, right?

Reply:
Yeah, we would need some sort of `AggregationKind` enum and a `Kind()` function on `Aggregation`.  It does feel nicer to have an enum than to have to do type assertions, since I can write a switch statement on it.  I don't think we would need to define `options` until an option is required, right?  Callers of aggregation.Drop() would continue to work if we changed `func Drop()` to `func Drop(opts ...DropOption)`.

From a user's perspective if they need to write a custom selector:

```golang
	var a aggregation.Aggregation
	switch ik {
	case SyncCounter, SyncUpDownCounter, AsyncCounter, AsyncUpDownCounter:
		a.Operation = aggregation.Sum{}
                ...
	case SyncHistogram:
		a.Operation = aggregation.ExplicitBucketHistogram{
			Boundaries:   []float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 1000},
			RecordMinMax: true,
                }
        }
```
vs

```golang
	switch ik {
	case SyncCounter, SyncUpDownCounter, AsyncCounter, AsyncUpDownCounter:
		return aggregation.Sum()
        ...
	case SyncHistogram:
		return aggregation.ExplicitBucketHistogram(
			aggregation.WithBoundaries([]float64{0, 5, 10, 25, 50, 75, 100, 250, 500, 1000}),
			aggregation.WithRecordMinMax(),
                }
        }
```

From a user's perspective if they are setting an aggregation on a View:

```golang
        view.New(view.WithSetAggregation(aggregation.Aggregation{Operation: aggregation.Drop}))
```
vs
```golang
        view.New(view.WithSetAggregation(aggregation.Drop()))
```

Overall, not requiring users to deal with Operation seems nice...

Reply:
> It would, so we would need to _only_ accept an `aggregation.Aggregation` in our interfaces, and call the Kind() method ourselves to determine the enum (i.e. we can't ever trust a raw aggregation.Kind provided by users).

Gotcha. I think this could work as an alternative to what is being presented here. I wonder though about getting around the type assertion with this. We would switch on the enum and then type assert the `Aggregation` to be one of a few values. I.e.

```go
func foo(a Aggregation) {
	switch a.Kind() {
	case ExplicitBucketHistogramAggregation:
		h := a.(ExplicitBucketHistogram)
		// Use  h.
	case /* ... */:
	}
}
```

I have pushed 33b2a8524826abefb12de212e1a0c97a0a132f3d to switch to the `Aggregation` interface. Let me know if you think we should still pursue the enum approach.

Also, for completeness ...

> I don't think we would need to define options until an option is required, right? Callers of aggregation.Drop() would continue to work if we changed `func Drop()` to func `Drop(opts ...DropOption)`.

Changing the function signature [is a breaking change](https://go.dev/blog/module-compatibility#adding-to-a-function) as they both represents a distinctly type.

Reply:
TIL, thanks.  Let's stick with the interface for now.